### PR TITLE
feat: add BinaryNumericFn for array arithmetic

### PR DESCRIPTION
**Pull request by danking**
I did not implement any binary numeric functions because it is not clear that there are any cases where we can out run decompression. Two run end arrays might be a happy path? Two dictionaries, maybe, if the dictionaries are much smaller than the decompressed arrays?

Binary scalar numeric functions are more obviously valuable: clickbench includes several uses of scalar add or subtract.

### CODE REVIEW

**Comment by gatesn** [comment]
Why ScalarNumeric instead of using is_constant as with all other compute functions?

**Comment by danking** [comment]
`subtract_scalar` was already present and this was the natural generalization.

We could remove `subtract_scalar` and friends and replace them with functions that create constant arrays of the right length and apply the binary operator and have the binary operators handle constants RHSes. What was the reasoning for `subtract_scalar`?

**Comment by danking** [comment]
Alright, scalar_subtract and scalar_numeric are completely gone. Four convenience functions survive: `sub_scalar`, etc. which delegate to `binary_numeric`.

**Comment by gatesn** [comment]
@danking none of our compute functions have internal casting.

Let's just check LHS and RHS are exactly equal (including null-ability) and fail if not. The caller has to decide casting rules, e.g. the compute engine, else it gets very confusing to keep track of coercion semantics

**Comment by danking** [comment]
@gatesn the old subtract_scalar casted the constant (which is what this PR preserves and extends to add, multiply, and divide), we seem to rely on this for shifting usize indices around. I could push the cast into the call sites though?

**Comment by gatesn** [comment]
Yeah I think the call site is best, albeit I bit more annoying

**Comment by danking** [comment]
@gatesn done.

File: vortex-array/src/array/chunked/mod.rs [hunk 1]
```diff
@@ -234,17 +234,25 @@ impl ValidityVTable<ChunkedArray> for ChunkedEncoding {
...
-        Ok(ChunkedArray::try_new(chunks, array.dtype().clone())?.into_array())
+        rhs: &ArrayData,
+        op: NumericOperator,
+    ) -> VortexResult<Option<ArrayData>> {
+        let mut start = 0;
+
+        let mut new_chunks = Vec::with_capacity(array.nchunks());
+        for chunk in array.chunks() {
+            let end = start + chunk.len();
+            new_chunks.push(binary_numeric(&chunk, &slice(rhs, start, end)?, op)?);
```
Comment (lwwmanning) [inline vortex-array/src/array/chunked/mod.rs:249]: AFAICT, `new_chunks` will be a Vec<Option<ArrayData>>... we probably want it to be forced to have non-empty chunks, right?

File: vortex-array/src/array/constant/compute/mod.rs [hunk 2]
```diff
@@ -48,6 +52,30 @@ impl ComputeVTable for ConstantEncoding {
...
+    ) -> VortexResult<Option<ArrayData>> {
+        let Some(rhs) = rhs.as_constant() else {
+            return Ok(None);
+        };
+
+        Ok(Some(
+            ConstantArray::new(
+                array
+                    .scalar()
+                    .as_primitive()
```
Comment (lwwmanning) [inline vortex-array/src/array/constant/compute/mod.rs]: might not be primitive?

File: vortex-array/src/compute/binary_numeric.rs [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+/// the RHS is not constant and expand to a full array.
+fn arrow_numeric(
+    lhs: ArrayData,
+    rhs: ArrayData,
+    operator: NumericOperator,
+) -> VortexResult<ArrayData> {
+    let nullable = lhs.dtype().is_nullable() || rhs.dtype().is_nullable();
+
+    let lhs = lhs.into_canonical()?.into_arrow()?;
+    let rhs = rhs.into_canonical()?.into_arrow()?;
```
Comment (lwwmanning) [inline vortex-array/src/compute/binary_numeric.rs]: check if rhs is constant (because maybe it is but lhs doesn't have binary_numeric_fn implemented)

Comment (gatesn) [reply vortex-array/src/compute/binary_numeric.rs]: You should do Datum::try_from instead which will use Arrow scalars where possible (see CompareFn)

File: vortex-array/src/array/chunked/mod.rs (hunk 1 above)
Comment (gatesn) [reply vortex-array/src/array/chunked/mod.rs:249]: Why would it be Vec<Option>?

File: vortex-array/src/array/constant/compute/mod.rs [hunk 4]
```diff
@@ -4,21 +4,25 @@ mod invert;
...
 use crate::array::constant::ConstantArray;
 use crate::array::ConstantEncoding;
 use crate::compute::{
-    BinaryBooleanFn, CompareFn, ComputeVTable, FilterFn, FilterMask, InvertFn, ScalarAtFn,
-    SearchSortedFn, SliceFn, TakeFn,
+    BinaryBooleanFn, BinaryNumericFn, CompareFn, ComputeVTable, FilterFn, FilterMask, InvertFn,
+    ScalarAtFn, SearchSortedFn, SliceFn, TakeFn,
 };
-use crate::{ArrayData, IntoArrayData};
+use crate::{ArrayData, ArrayLen as _, IntoArrayData};
```
Comment (gatesn) [inline vortex-array/src/array/constant/compute/mod.rs]: It's you!! You're the IDE that inserts `as _`

File: vortex-array/src/array/constant/compute/mod.rs [hunk 5]
```diff
@@ -48,6 +52,30 @@ impl ComputeVTable for ConstantEncoding {
     }
 }
 
+impl BinaryNumericFn<ConstantArray> for ConstantEncoding {
```
Comment (gatesn) [inline vortex-array/src/array/constant/compute/mod.rs]: I know it's annoying, but where we've already converted `compute` to be a module, can we try to put each compute function in its own file?

File: vortex-array/src/array/constant/compute/mod.rs (hunk 2 above)
Comment (gatesn) [reply vortex-array/src/array/constant/compute/mod.rs]: Should be for numeric compute. Implementations of compute functions are allowed to make assumptions that are validated by the entry point function (else everyone has to check index bounds and dtypes everywhere etc etc)

File: vortex-array/src/array/null/compute.rs [hunk 6]
```diff
@@ -22,6 +26,18 @@ impl ComputeVTable for NullEncoding {
     }
 }
 
+impl BinaryNumericFn<NullArray> for NullEncoding {
```
Comment (gatesn) [inline vortex-array/src/array/null/compute.rs:29]: I don't think Null arrays support numeric operations. This isn't nullable int, this is literally only null. Let's remove it

File: vortex-array/src/compute/binary_numeric.rs [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+        || lhs.dtype() != rhs.dtype()
+    {
+        vortex_bail!(
+            "Numeric operations are only supported on two arrays sharing the same primitive-type: {} {}",
+            lhs.dtype(),
+            rhs.dtype()
+        )
+    }
+
+    // Check if LHS supports the operation directly.
```
Comment (gatesn) [inline vortex-array/src/compute/binary_numeric.rs:117]: You should invert LHS and RHS for commutative operators (add, mul) and check if RHS supports LHS.

File: vortex-array/src/encoding/mod.rs [hunk 8]
```diff
@@ -67,6 +69,19 @@ pub trait Encoding: 'static {
     type Metadata: ArrayMetadata;
 }
 
+pub fn downcast_array<E: Encoding>(array: &ArrayData) -> VortexResult<(&E::Array, &E)>
```
Comment (gatesn) [inline vortex-array/src/encoding/mod.rs]: Can this be called `downcast_array_ref`? Since we have a non-ref'd version of this too.

Also maybe this should happen in a FLUP PR that converts everywhere.

File: vortex-dtype/src/ptype.rs [hunk 9]
```diff
@@ -71,6 +71,18 @@ pub trait NativePType:
     /// Compare another instance of this type to `self`, providing a total ordering
     fn total_compare(self, other: Self) -> Ordering;
 
+    /// Addition, safe on uninitialized data.
```
Comment (gatesn) [inline vortex-dtype/src/ptype.rs]: I don't really know what this function name means, and the comment doesn't really help?

File: vortex-dtype/src/ptype.rs [hunk 10]
```diff
@@ -88,6 +100,22 @@ macro_rules! native_ptype {
                 self.cmp(&other)
             }
 
+            fn maybe_null_add(self, other: Self) -> Self {
```
Comment (gatesn) [inline vortex-dtype/src/ptype.rs]: I don't think we should define these operators, and they _almost certainly_ should not be wrapping

File: vortex-scalar/src/primitive.rs [hunk 11]
```diff
@@ -270,3 +272,47 @@ impl From<usize> for Scalar {
...
 }
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum NumericOperator {
+    Add,
+    Sub,
+    Mul,
+    Div,
+    // Missing from arrow-rs:
+    // Min,
```
Comment (gatesn) [inline vortex-scalar/src/primitive.rs:283]: Some of these aren't binary operators

File: vortex-scalar/src/primitive.rs [hunk 12]
```diff
@@ -270,3 +272,47 @@ impl From<usize> for Scalar {
...
+    // RDiv,
+}
+
+impl PrimitiveScalar<'_> {
+    pub fn numeric_operator(
+        self,
+        other: PrimitiveScalar<'_>,
+        op: NumericOperator,
+    ) -> VortexResult<Scalar> {
+        if !self.dtype().eq_ignore_nullability(other.dtype()) {
```
Comment (gatesn) [inline vortex-scalar/src/primitive.rs:301]: Why are we ignoring nullability? Not saying we shouldn't be (although maybe we shouldn't be), but if we do this it should be commented.

File: vortex-scalar/src/primitive.rs [hunk 13]
```diff
@@ -270,3 +272,47 @@ impl From<usize> for Scalar {
...
+            vortex_bail!("types must match: {} {}", self.dtype(), other.dtype());
+        }
+
+        Ok(match_each_native_ptype!(self.ptype(), |$P| {
+            let lhs = self.typed_value::<$P>();
+            let rhs = other.typed_value::<$P>();
+            let nullability = self.dtype().nullability();
+            lhs.zip(rhs).map(|(lhs, rhs)| {
```
Comment (gatesn) [inline vortex-scalar/src/primitive.rs]: Zipping optionals is quite hard to parse... I guess it means if either side is None then the result is an empty iterator?

File: vortex-scalar/src/primitive.rs [hunk 14]
```diff
@@ -270,3 +272,47 @@ impl From<usize> for Scalar {
...
+                match op {
+                    NumericOperator::Add => Scalar::primitive(lhs + rhs, nullability),
+                    NumericOperator::Sub => Scalar::primitive(lhs - rhs, nullability),
+                    NumericOperator::Mul => Scalar::primitive(lhs * rhs, nullability),
+                    NumericOperator::Div => Scalar::primitive(lhs / rhs, nullability),
+                }
+            }).unwrap_or_else(|| Scalar::null(self.dtype().clone()))
```
Comment (gatesn) [inline vortex-scalar/src/primitive.rs]: Because you checked for dtype_eq_ignore_nullability, you may now end up trying to construct a non-nullable null scalar right?

File: vortex-scalar/src/primitive.rs [hunk 15]
```diff
@@ -270,3 +272,47 @@ impl From<usize> for Scalar {
...
+    // Pow,
+    // In IEEE-754 operations involving NaN are non-commutative wrt NaN payload/quietness
+    // RAdd,
+    // RSub,
+    // RMul,
+    // RDiv,
+}
+
+impl PrimitiveScalar<'_> {
+    pub fn numeric_operator(
```
Comment (gatesn) [inline vortex-scalar/src/primitive.rs]: Can we call this `checked_numeric_operator` and return `VortexResult<Option<Scalar>>`, then use checked_add, etc from num-traits to perform the operators.

**Review by gatesn — CHANGES_REQUESTED** [review]
Sorry it's taken me a while to get to this

File: vortex-array/src/array/chunked/mod.rs (hunk 1 above)
Comment (danking) [reply vortex-array/src/array/chunked/mod.rs:249]: The `binary_numeric` top-level function converts to arrow if the LHS lacks a specialization, so there are no Nones in this vec.

In this case, ChunkedArray should never return None because that would force copies to create a contiguous array.

In the dict and runend cases, we _could_ check if the child in question supports this operation; however, it seems like, in both dict and runend, it's always preferable to decompress and apply the operation rather than for dict or runend to decompress and then apply the operation.

File: vortex-array/src/array/constant/compute/mod.rs (hunk 5 above)
Comment (danking) [reply vortex-array/src/array/constant/compute/mod.rs]: Done. Should I also move scalar_at, take, slice, and filter to new files or is there a line count threshold at which a new file is justified?

File: vortex-array/src/array/constant/compute/mod.rs (hunk 2 above)
Comment (danking) [reply vortex-array/src/array/constant/compute/mod.rs]: Indeed, [the top-level binary_numeric verifies dtypes](https://github.com/spiraldb/vortex/blob/dk/arithmetic/vortex-array/src/compute/binary_numeric.rs#L106) should verify the dtype of array and rhs.

File: vortex-array/src/array/null/compute.rs (hunk 6 above)
Comment (danking) [reply vortex-array/src/array/null/compute.rs:29]: Ah, I was mistaken. Done.

File: vortex-array/src/compute/binary_numeric.rs (hunk 7 above)
Comment (danking) [reply vortex-array/src/compute/binary_numeric.rs:117]: Agreed, though I would need to branch here on the dtype to ensure I don't do that to floats _or_ we could decide that we ignore IEEE 754's rules about the non-commutativity of NaN. It seemed like eventually adding `R<operator>` involved fewer special cases.

File: vortex-array/src/compute/binary_numeric.rs (hunk 3 above)
Comment (danking) [reply vortex-array/src/compute/binary_numeric.rs]: Now using `Datum::try_from`.

File: vortex-array/src/encoding/mod.rs (hunk 8 above)
Comment (danking) [reply vortex-array/src/encoding/mod.rs]: Changed. I'll follow up with a PR to do it everywhere.

File: vortex-dtype/src/ptype.rs (hunk 9 above)
Comment (danking) [reply vortex-dtype/src/ptype.rs]: Regarding this comment [and the one below](https://github.com/spiraldb/vortex/pull/1640/files/704fceb9c932cca8b781d5c42d78dd37642737b7#r1885025131), the short answer is: I actually do not need these so I have removed them.

The longer answer is: in develop, we [use wrapping sub](https://github.com/spiraldb/vortex/blob/develop/vortex-array/src/array/primitive/compute/subtract_scalar.rs#L64-L68) to implement `subtract_scalar` when the `maybe_null_slice` contains nulls. Perhaps I am wrong, but assumed that null entries are not necessary initialized memory. If they're not necessarily initialized, performing arithmetic could overflow which will panic in debug mode. I assume we explicitly ask for wrapping to ensure debug and release have the same behavior (albeit on data that should never affect the program's semantics).

I avoided this whole mess by delegating arithmetic for `PrimitiveArray` to Arrow.

File: vortex-dtype/src/ptype.rs (hunk 10 above)
Comment (danking) [reply vortex-dtype/src/ptype.rs]: See reply to [this comment](https://github.com/spiraldb/vortex/pull/1640/files/704fceb9c932cca8b781d5c42d78dd37642737b7#r1885024953)

File: vortex-scalar/src/primitive.rs (hunk 11 above)
Comment (danking) [reply vortex-scalar/src/primitive.rs:283]: I renamed the enum to BinaryNumericOperator and removed Sqrt. All the others have a sensible point wise interpretation on two arrays.

File: vortex-scalar/src/primitive.rs (hunk 15 above)
Comment (danking) [reply vortex-scalar/src/primitive.rs]: Done.

File: vortex-scalar/src/primitive.rs (hunk 12 above)
Comment (danking) [reply vortex-scalar/src/primitive.rs:301]: Supporting different nullabilities isn't difficult and does not seem to me likely to affect speed much since we're already working with `Scalar` rather than primitives. Compare works similarly. What kind of comment are you looking for?

File: vortex-scalar/src/primitive.rs (hunk 13 above)
Comment (danking) [reply vortex-scalar/src/primitive.rs]: If either is None, the result is None otherwise `Some((left, right))`. I actually (almost always) prefer match but the team seems to have heterogenous opinions on combinators versus match.

I'll switch to a `match`.

File: vortex-scalar/src/primitive.rs (hunk 14 above)
Comment (danking) [reply vortex-scalar/src/primitive.rs]: Now using `as_nullable()`.

**Review by danking — COMMENTED** [review]
OK, I resolved all the threads that I think are uncontroversially resolved. I think the unresolved ones still need confirmation or more discussion.

File: vortex-array/src/array/chunked/mod.rs (hunk 1 above)
Comment (gatesn) [reply vortex-array/src/array/chunked/mod.rs:249]: Yeah chunked is weird... We've found in a few places that canonicalize is faster than the overhead of chunks. We should think about where/how we decide to concatenate these things.

File: vortex-scalar/src/primitive.rs (hunk 12 above)
Comment (gatesn) [reply vortex-scalar/src/primitive.rs:301]: Only because we of the general approach in Vortex that a compute function should never perform type coercion.

File: vortex-scalar/src/primitive.rs (hunk 14 above)
Comment (gatesn) [reply vortex-scalar/src/primitive.rs]: Are you?

File: vortex-array/src/compute/binary_numeric.rs (hunk 7 above)
Comment (lwwmanning) [reply vortex-array/src/compute/binary_numeric.rs:117]: IEEE 754 enforces commutativity unless both operands are NaN... in which case, it allows implementations to propagate *either* NaN operand. So we should just treat float addition/multiplication as commutative

**Review by lwwmanning — APPROVED** [review]

File: vortex-scalar/src/primitive.rs (hunk 14 above)
Comment (danking) [reply vortex-scalar/src/primitive.rs]: Yeah, the `address comments` commit [did it here](https://github.com/spiraldb/vortex/pull/1640/commits/4ebb7edad2b454874df0aced585c204bd71e73fd#diff-8f1cbe207a4076299ad7913ac7d7d0db5fd2e99d1a689f984fcedfd1b91208faR315-R333).

File: vortex-scalar/src/primitive.rs [hunk 16]
```diff
@@ -270,3 +272,70 @@ impl From<usize> for Scalar {
...
+
+        let nullability = self.dtype().nullability();
+
+        Ok(match_each_native_ptype!(
+            self.ptype(),
+            integral: |$P| {
+                let lhs = self.typed_value::<$P>();
+                let rhs = other.typed_value::<$P>();
+                match (lhs, rhs) {
+                    (_, None) | (None, _) => Some(Scalar::null(self.dtype().clone().as_nullable())),
```
Comment (gatesn) [inline vortex-scalar/src/primitive.rs]: This is the bug I think is still here. If `(_, None)` is true, and lhs is non-nullable, then you're going to try to create a Scalar::null that's non-nullable

**Review by gatesn — CHANGES_REQUESTED** [review]

File: vortex-scalar/src/primitive.rs [hunk 17]
```diff
@@ -270,3 +272,70 @@ impl From<usize> for Scalar {
...
+    pub fn checked_numeric_operator(
+        self,
+        other: PrimitiveScalar<'_>,
+        op: BinaryNumericOperator,
+    ) -> VortexResult<Option<Scalar>> {
+        if !self.dtype().eq_ignore_nullability(other.dtype()) {
+            vortex_bail!("types must match: {} {}", self.dtype(), other.dtype());
+        }
+
+        let nullability = self.dtype().nullability();
```
Comment (gatesn) [inline vortex-scalar/src/primitive.rs]: I also think this should be `let nullability = self.dtype.is_nullable() || other.dtype.is_nullable()`

Comment (danking) [reply vortex-scalar/src/primitive.rs]: Done.

File: vortex-scalar/src/primitive.rs (hunk 16 above)
Comment (danking) [reply vortex-scalar/src/primitive.rs]: This case is correct (b/c of the as_nullable) but now every case uses the same (least viable) nullability.

**Review by gatesn — APPROVED** [review]
